### DISCUSSION TITLE
emqx_plugin_template Customization for Message Hooks

### DISCUSSION THREAD

**Discussion by SSaiPrasad7** [emqx/emqx#4543] [Q&A] [answered]
Hello @zmstone 
I hope everything is fine with you. I had added `emqx-plugin-template` plugin to EMQX server that you told earlier. The plugin works fine. I would like to know how to customize the plugin for `msg.payload`?
Thanks for replies for previous discussions. I hope you could resolve this.
Awaiting for your reply. 
All the best for the future projects.

**Comment by zmstone** [comment, marked as answer]
Hi @saiprasadsamudrala 
Could you be more specific about what you want to customise with `msg.payload`?

**Reply by SSaiPrasad7** [reply]
![message_hooks](https://user-images.githubusercontent.com/36000620/114385909-30300400-9bae-11eb-8b73-44548a515933.png)
I am unable to get the payload on the console by using `on_message_publish` hook in `emqx_plugin_template` as you see in the highlighted message on the image.

**Reply by zmstone** [reply]
If you haven’t changed the code, it’s 

```
on_message_publish(Message, _Env) ->
    io:format("Publish ~s~n", [emqx_message:format(Message)]),
    {ok, Message}
```

Which calls this format function:
https://github.com/emqx/emqx/blob/e2c2f39bc2737b016c63d63bcabd58068e3b1bc8/src/emqx_message.erl#L344-L345
it ignores the payload.

if you change the callback to

```
on_message_publish(Message, _Env) ->
    io:format("Publish ~p~n", [Message]),
    {ok, Message}
```

You’ll see the payload.

**Reply by SSaiPrasad7** [reply]
@zmstone,I had made change and reloaded the plugin but it remains same as earlier.

![emqx_hook](https://user-images.githubusercontent.com/36000620/114535029-3f29bb80-9c6d-11eb-9f83-08252b40570f.png)

**Reply by zmstone** [reply]
please try `make clean`
and compile then run again.

**Reply by SSaiPrasad7** [reply]
===> Compiling emqx_stomp
===> Compiling emqx_web_hook
===> Compiling emqx_dashboard
===> Compiling emqx_modules
===> Compiling emqx_telemetry
===> [relup_helper] running gen_appups
===> Uncaught error in rebar_core. Run with DIAGNOSTIC=1 to see stacktrace or consult rebar3.crashdump
===> When submitting a bug report, please include the output of `rebar3 report "your command"`
make: *** [Makefile:77: emqx] Error 1


I got an error after running `make clean` and then `make`

**Reply by zmstone** [reply]
please try this:
```
export DIAGNOSTIC=1
make
```

**Reply by SSaiPrasad7** [reply]
I tried with the commands still I have an error on console.
Error Message:
===> When submitting a bug report, please include the output of `rebar3 report "your command"`
make: *** [Makefile:77: emqx] Error 1

**Reply by SSaiPrasad7** [reply]
@maneetgoyal

**Reply by zmstone** [reply]
```
export DIAGNOSTIC=1
make | tee failed.log
```

this will dump the detailed reasons why it failed to the file failed.log
Please create a github gist, upload the file so we can inspect it

**Reply by SSaiPrasad7** [reply]
https://gist.github.com/saiprasadsamudrala/969f640106aacb79ccd090fdc432bcc3

**Reply by zmstone** [reply]
@saiprasadsamudrala 
have you set `export EMQX_EXTRA_PLUGINS=emqx_plugin_template` ?

**Reply by SSaiPrasad7** [reply]
No.
Sorry, I did now and it is working now.
Thanks for your support .I am very happy with responses I got from your side.
